lib/ClangImporter/ClangDerivedConformances.cpp

```diff
@@ -1364,3 +1364,218 @@ void swift::conformToCxxSpanIfNeeded(ClangImporter::Implementation &impl,
+      rhsParamId, rhsTy, rhsTyInfo, clang::StorageClass::SC_None, nullptr);
+  auto rhsParamRefExpr = new (clangCtx) clang::DeclRefExpr(
+      clangCtx, rhsParamDecl, false, rhsTy, clang::ExprValueKind::VK_LValue,
+      clang::SourceLocation());
+
+  opDecl->setParams({lhsParamDecl, rhsParamDecl});
+
+  // looking for the constructor for the function object
+  clang::CXXConstructorDecl *ctorDecl = nullptr;
+  for (clang::CXXConstructorDecl *candidate : funcDecl->ctors()) {
+    if (candidate->param_size() == 0) {
+      ctorDecl = candidate;
+      break;
+    }
+  }
+
+  if (!ctorDecl)
+    return false;
+
+  clang::QualType funcTy = clangCtx.getRecordType(funcDecl);
+  clang::SourceLocation funcLoc = funcDecl->getLocation();
+
+  clang::ExprResult synthCtorExprResult = clangSema.BuildCXXConstructExpr(
+      funcLoc, funcTy, ctorDecl,
+      /*Elidable=*/false, {},
+      /*HadMultipleCandidates=*/false,
+      /*IsListInitialization=*/false,
+      /*IsStdInitListInitialization=*/false,
+      /*RequiresZeroInit=*/false, clang::CXXConstructionKind::Complete,
+      clang::SourceRange(funcLoc, funcLoc));
+
+  ASSERT(!synthCtorExprResult.isInvalid() &&
+         "Unable to synthesize constructor expression for std::equal_to");
+  clang::Expr *synthCtorExpr = synthCtorExprResult.get();
+
+  SmallVector<clang::Expr *, 2> args = {lhsParamRefExpr, rhsParamRefExpr};
+  auto call =
+      clangSema.BuildCallExpr(nullptr, synthCtorExpr, clang::SourceLocation(),
+                              args, clang::SourceLocation());
+
+  if (!call.isUsable())
+    return false;
+
+  auto opBody = clang::ReturnStmt::Create(clangCtx, clang::SourceLocation(),
+                                          call.get(), nullptr);
+  opDecl->setBody(opBody);
+
+  impl.synthesizedAndAlwaysVisibleDecls.insert(opDecl);
+  auto lookupTable1 = impl.findLookupTable(classDecl);
+  addEntryToLookupTable(*lookupTable1, opDecl, impl.getNameImporter());
+  auto owningModule = impl.getClangOwningModule(classDecl);
+  auto lookupTable2 = impl.findLookupTable(owningModule);
+  if (lookupTable1 != lookupTable2)
+    addEntryToLookupTable(*lookupTable2, opDecl, impl.getNameImporter());
+  return true;
+}
+
+static clang::ClassTemplateSpecializationDecl *
+lookupAndSpecializeFunctionObject(ClangImporter::Implementation &impl,
+                                  const clang::CXXRecordDecl *clangDecl,
+                                  std::string name) {
+  ASSERT(clangDecl);
+  clang::ASTContext &Ctx = impl.getClangASTContext();
+  clang::Sema &clangSema = impl.getClangSema();
+
+  clang::NamespaceDecl *stdNS = clangSema.getStdNamespace();
+  // maybe the C++ header didn't include standard library?
+  if (!stdNS)
+    return nullptr;
+
+  clang::DeclarationName declName =
+      Ctx.DeclarationNames.getIdentifier(&Ctx.Idents.get(name));
+
+  clang::ClassTemplateDecl *tmplDecl =
+      stdNS->lookup(declName).find_first<clang::ClassTemplateDecl>();
+
+  if (!tmplDecl)
+    return nullptr;
```

Comment:
Does this lookup need to be repeated (twice) upon every import? This seems like the kind of thing that should be memoized by `ClangImporter`, somewhere, since it is not dependent on the `clangDecl`, which is only used later to find the specialization.

Reply:
I'm confused by the "(twice)" part, but I assumed you mean the `ClassTemplateDecl` for `name`? I think yes it should be memorized by the importer.

Though I'm not familiar with that part of the code structure. Which class would be the best place to store these lookup information? Is it the `ClangImporter::Implementation` class?

Reply:
> I'm confused by the "(twice)" part, but I assumed you mean the ClassTemplateDecl for name? I think yes it should be memorized by the importer.

Yeah, that's what I meant. I'm not sure if there's any cacheing on the Clang side that makes `stdNS->lookup()` efficient but I don't think it should happen `2 * N` times if we import `N` structs (once for `std::hash` and once for `std::equal_to`).

And yeah, the "easiest" thing to do is to stuff it in `ClangImporter::Implementation` as a member variable, and you can lazily initialize it or something like that. Though that struct is a bit cluttered, so adding another field might be bad from an organization standpoint. @egorzhdan what are your thoughts here?